### PR TITLE
Implement internal connection pools for the read-write plugin

### PR DESCRIPTION
**Pull request by congoamz**
Add functionality to use internal connection pools to establish connections. The motivation for using pools comes from the read-write plugin's need to regularly establish new connections. Using pools will increase performance and decrease demands on the database by allowing connections to be re-used.

### By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: wrapper/src/main/java/software/amazon/jdbc/ConnectionProvider.java [hunk 1]
```diff
@@ -49,4 +62,9 @@ Connection connect(
    */
   Connection connect(@NonNull String url, @NonNull Properties props)
       throws SQLException; // TODO: do we need this method?
+
+  /**
+   * Release any resources used by this ConnectionProvider
+   */
+  void releaseResources();
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/ConnectionProvider.java:69]: I would not extend interface with this method but allow each particular class to implement CanReleaseResources if needed.

File: wrapper/src/main/java/software/amazon/jdbc/ConnectionProviderManager.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+ */
+
+import java.util.Properties;
+import java.util.concurrent.locks.ReentrantLock;
+
+public class ConnectionProviderManager {
+
+  private static ConnectionProvider connProvider = null;
+  private ConnectionProvider defaultProvider;
+  private static final ReentrantLock lock = new ReentrantLock();
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/ConnectionProviderManager.java:24]: In this particular case I think ReentrantReadWriteLock could be better. We shouldn't care about it now. It's a POC.

File: wrapper/src/main/java/software/amazon/jdbc/ConnectionProviderManager.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    return defaultProvider;
+  }
+
+  public static void reset() {
+    lock.lock();
+    connProvider = null;
+    lock.unlock();
+  }
+
+  public static void releaseResources() {
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/ConnectionProviderManager.java:62]: It shouldn't be static if it's a part of an interface.

File: wrapper/src/main/java/software/amazon/jdbc/HikariPooledConnectionProvider.java [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+  static final String PG_DRIVER_PROTOCOL = "jdbc:postgresql://";
+  static final String MYSQL_DRIVER_PROTOCOL = "jdbc:mysql://";
+  static final String MARIADB_DRIVER_PROTOCOL = "jdbc:mariadb://";
+  static final String DEFAULT_PG_DS = "org.postgresql.ds.PGSimpleDataSource";
+  static final String DEFAULT_MYSQL_DS = "com.mysql.cj.jdbc.MysqlDataSource";
+  static final String DEFAULT_MARIADB_DS = "org.mariadb.jdbc.MariaDbDataSource";
+
+  private static final RdsUtils rdsUtils = new RdsUtils();
+  private static final Map<String, HikariDataSource> databasePools = new ConcurrentHashMap<>();
+  private static HikariPoolConfigurator poolConfigurator;
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/HikariPooledConnectionProvider.java:42]: why static?

File: wrapper/src/main/java/software/amazon/jdbc/HikariPooledConnectionProvider.java [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+    }
+    if (password != null) {
+      config.setPassword(password);
+    }
+
+    if (HostRole.READER.equals(hostSpec.getRole())) {
+      config.setReadOnly(true);
+    }
+
+    switch (protocol) {
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/HikariPooledConnectionProvider.java:85]: I'd implement a base abstract class with 3 subclasses for PG, mysql and maridb. It's a classic case for OOP.

File: wrapper/src/main/java/software/amazon/jdbc/HikariPooledConnectionProvider.java [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+    String db = PropertyDefinition.DATABASE.getString(props);
+    if (!StringUtils.isNullOrEmpty(dbPropertyName) && !StringUtils.isNullOrEmpty(db)) {
+      config.addDataSourceProperty(dbPropertyName, db);
+    }
+  }
+
+  @Override
+  public Connection connect(
+      @NonNull String url, @NonNull Properties props) throws SQLException {
+    // TODO: Is this method ever called? It seems to only be referenced by tests/benchmarks. If
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/HikariPooledConnectionProvider.java:120]: Agree, it don't recall it's ever used. Might consider to deprecate it.

File: wrapper/src/main/java/software/amazon/jdbc/HikariPooledConnectionProvider.java [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+import java.util.Map;
+import java.util.Properties;
+import java.util.concurrent.ConcurrentHashMap;
+import org.checkerframework.checker.nullness.qual.NonNull;
+import software.amazon.jdbc.util.HikariCPSQLException;
+import software.amazon.jdbc.util.RdsUrlType;
+import software.amazon.jdbc.util.RdsUtils;
+import software.amazon.jdbc.util.StringUtils;
+
+public class HikariPooledConnectionProvider implements PooledConnectionProvider {
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/HikariPooledConnectionProvider.java:32]: implements PooledConnectionProvider, CanReleaseResources
?

File: .editorconfig [hunk 8]
```diff
@@ -43,7 +43,7 @@ ij_any_spaces_around_shift_operators = true
 ij_continuation_indent_size = 4
 ij_java_align_multiline_parameters = false
 ij_java_if_brace_force = always
-ij_java_indent_case_from_switch = false
+ij_java_indent_case_from_switch = true
```
Comment (congoamz) [inline .editorconfig:46]: This setting was not aligned with our checkstyle config

File: benchmarks/src/jmh/java/software/amazon/jdbc/benchmarks/testplugin/BenchmarkPlugin.java [hunk 9]
```diff
@@ -65,6 +64,24 @@ public Connection connect(String driverProtocol, HostSpec hostSpec, Properties p
     return connectFunc.call();
   }
 
+  @Override
+  public Connection forceConnect(String driverProtocol, HostSpec hostSpec, Properties props,
+      boolean isInitialConnection, JdbcCallable<Connection, SQLException> forceConnectFunc)
+      throws SQLException {
+    LOGGER.finer(() -> String.format("forceConnect=''%s''", driverProtocol));
```
Comment (karenc-bq) [inline benchmarks/src/jmh/java/software/amazon/jdbc/benchmarks/testplugin/BenchmarkPlugin.java:71]: Should this be driverprotocol here? Do we want to use the hostname for more information?

File: wrapper/src/main/java/software/amazon/jdbc/HikariPooledConnectionProvider.java [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+
+  @Override
+  public void releaseResources() {
+    databasePools.forEach((String url, HikariDataSource ds) -> ds.close());
+    databasePools.clear();
+  }
+
+  abstract String getDataSourceClassName();
+
+  private void setConnectionProperties(
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/HikariPooledConnectionProvider.java:93]: Probably protected would be better. There's a big chance that we might need to override it in subclasses. Just a suggestion.

File: wrapper/src/main/java/software/amazon/jdbc/MariaDBHikariPooledConnectionProvider.java [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+package software.amazon.jdbc;
+
+import software.amazon.jdbc.util.StringUtils;
+
+public class MariaDBHikariPooledConnectionProvider extends HikariPooledConnectionProvider {
+  private static final String DEFAULT_MARIADB_DATA_SOURCE = "com.mysql.cj.jdbc.MysqlDataSource";
+  private String dataSourceClassName;
+
+  public MariaDBHikariPooledConnectionProvider(
+      HikariPoolConfigurator configurator, String dataSourceClassName) {
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/MariaDBHikariPooledConnectionProvider.java:26]: I'm not sure about dataSourceClassName as a parameter. What is it for?

File: wrapper/src/main/java/software/amazon/jdbc/MariaDBHikariPooledConnectionProvider.java [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+    this.dataSourceClassName = dataSourceClassName;
+  }
+
+  public MariaDBHikariPooledConnectionProvider(HikariPoolConfigurator configurator) {
+    super(configurator);
+  }
+
+  @Override
+  String getDataSourceClassName() {
+    return StringUtils.isNullOrEmpty(this.dataSourceClassName)
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/MariaDBHikariPooledConnectionProvider.java:37]: return DEFAULT_MARIADB_DATA_SOURCE
?

File: wrapper/src/main/java/software/amazon/jdbc/RandomHostSelector.java [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+
+  @Override
+  public HostSpec getHost(List<HostSpec> hosts, HostRole role) throws SQLException {
+    List<HostSpec> eligibleHosts = hosts.stream()
+        .filter(hostSpec -> role.equals(hostSpec.getRole())).collect(Collectors.toList());
+    if (eligibleHosts.size() == 0) {
+      throw new SQLException(Messages.get("RandomHostSelector.noHostsMatchingRole", new Object[]{role}));
+    }
+
+    int randomIndex = (int) (Math.random() * eligibleHosts.size());
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/RandomHostSelector.java:34]: int randomIndex = new Random().nextInt(eligibleHosts.size())
?

File: wrapper/src/main/java/software/amazon/jdbc/plugin/DefaultConnectionPlugin.java [hunk 14]
```diff
@@ -137,16 +139,60 @@ public Connection connect(
...
+  }
+
+  @Override
+  public HostSpec getHostSpecByStrategy(HostRole role, String strategy,
+      JdbcCallable<HostSpec, SQLException> getHostSpecByStrategyFunc)
+      throws SQLException {
+    Connection conn = this.pluginService.getCurrentConnection();
+    try {
+      if (conn != null && !conn.isClosed()) {
+        this.pluginService.refreshHostList(conn);
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/plugin/DefaultConnectionPlugin.java:181]: I wouldn't update host list here.

File: wrapper/src/main/java/software/amazon/jdbc/ConnectionProviderManager.java [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+    try {
+      ConnectionProviderManager.connProvider = connProvider;
+    } finally {
+      connProviderLock.writeLock().unlock();
+    }
+  }
+
+  public ConnectionProvider getConnectionProvider(
+      String driverProtocol, HostSpec host, Properties props) {
+    connProviderLock.readLock().lock();
```
Comment (davecramer) [inline wrapper/src/main/java/software/amazon/jdbc/ConnectionProviderManager.java:46]: Have to use the same lock to read and write

File: docs/using-the-jdbc-driver/using-plugins/UsingTheReadWriteSplittingPlugin.md [hunk 16]
```diff
@@ -33,6 +33,32 @@ String connectionUrl = "jdbc:aws-wrapper:mysql://writer-instance-1.com,reader-in
 
 Additionally, you should avoid using the Aurora host list plugin and the failover plugin in this scenario, as they are designed to specifically operate against Aurora databases.
 
+### Internal connection pooling
+
+Whenever `setReadOnly(true)` is first called on a `Connection` object, the read-write plugin will internally open a new physical connection to a reader. After this first call, the physical reader connection will be cached for the given `Connection`. Future calls to `setReadOnly `on the same `Connection` object will not require opening a new physical connection. However, calling `setReadOnly(true)` for the first time on a new `Connection` object will require the plugin to establish another new physical connection to a reader. If your application frequently establishes new read-only connections, you can enable internal connection pooling to improve performance. When enabled, the wrapper driver will maintain an internal connection pool for each instance in the cluster. This allows the read-write plugin to reuse reader connections that were established by previous `Connection` objects.
```
Comment (davecramer) [inline docs/using-the-jdbc-driver/using-plugins/UsingTheReadWriteSplittingPlugin.md:38]: However, calling `setReadOnly(true)` for the first time on a new `Connection` object will require the plugin to establish another new physical connection to a reader. 

Seems to just repeat the information above.

**Review by karenc-bq — APPROVED** [review]

**Pull request merged**